csp_report.php

```diff
@@ -76,26 +84,65 @@
 $table->sortable(true, 'failcounter', SORT_DESC);
 $table->define_columns(array(
     'failcounter',
-    'documenturi',
-    'blockeduri',
     'violateddirective',
-    'timecreated',
+    'blockeduri',
+    'highestviolaters',
     'timeupdated',
     'action',
 ));
 $table->define_headers(array(
     $failcounter,
-    $documenturi,
-    $blockeduri,
     $violateddirective,
-    $timecreated,
+    $blockeduri,
+    $highestviolaters,
     $timeupdated,
     $action,
 ));
 
-$fields = 'id, sha1hash, documenturi, blockeduri, violateddirective, failcounter, timecreated, timeupdated';
-$from = '{local_csp}';
-$where = '1 = 1';
+// If user has clicked on a violation to view all violation entries
+if (($viewviolationclass = optional_param('viewviolationclass', false, PARAM_TEXT)) !== false && confirm_sesskey()) {
+    $fields = 'id, sha1hash, blockeduri, violateddirective, failcounter, timeupdated, documenturi';
+    $from = "{local_csp}";
+    $where = "blockeduri = '$viewviolationclass'";
+
+    // Redefine columns to display Violation source
+    $table->define_columns(array(
+        'failcounter',
+        'violateddirective',
+        'blockeduri',
+        'documenturi',
+        'timeupdated',
+        'action',
+    ));
+    $table->define_headers(array(
+        $failcounter,
+        $violateddirective,
+        $blockeduri,
+        $documenturi,
+        $timeupdated,
+        $action,
+    ));
+
+} else {
+    $fields = 'id, sha1hash, blockeduri, violateddirective, failcounter, timeupdated';
+    // Select the first blockedURI of a type, and collapse the rest while summing failcounter
+    // Then grab other fields from the table where id is the selected collapsed ID
+    $from = "(SELECT A.id,
```

Comment:
this sql seems more complicated that is needs to be?

I suspect you don't need the subquery and can do it all as a single group by